src/gui/src/staGui.cpp

```diff
@@ -785,4 +873,377 @@ void TimingPathRenderer::highlightStage(gui::Painter& painter,
+  }
+
+  // draw instances
+  std::set<odb::dbInst*> instances;
+  for (const auto& [level, pins] : map_) {
+    for (const auto& pin : pins) {
+      if (pin->isPinITerm()) {
+        odb::dbInst* inst = pin->getPinAsITerm()->getInst();
+
+        if (inst != nullptr) {
+          instances.insert(inst);
+        }
+      }
+    }
+  }
+  auto* inst_descriptor = Gui::get()->getDescriptor<odb::dbInst*>();
+  painter.setPen(inst_color_, true);
+  painter.setBrush(inst_color_);
+  for (auto* inst : instances) {
+    inst_descriptor->highlight(inst, painter);
+  }
+
+  // draw timing connections
+  const double timing_range = max_timing_ - min_timing_;
+
+  auto timingToRatio = [this, timing_range](const TimingPathNode* node) {
+    if (timing_range == 0.0) {
+      return 1.0;
+    }
+    double value = 0.0;
+    if (node->hasValues()) {
+      value = 1.0 - (node->getPathSlack() - min_timing_) / timing_range;
+    }
+    return value;
+  };
+
+  const int line_width = 2; // 2 pixels
+  auto* iterm_descriptor = Gui::get()->getDescriptor<odb::dbITerm*>();
+  auto* bterm_descriptor = Gui::get()->getDescriptor<odb::dbBTerm*>();
+  for (const auto& [level, pins] : map_) {
+    for (const auto& pin : pins) {
+      const auto color = color_generator_.getColor(timingToRatio(pin.get()), 255);
+      painter.setPen(color, true);
+      painter.setBrush(color);
+
+      if (pin->isPinITerm()) {
+        iterm_descriptor->highlight(pin->getPinAsITerm(), painter);
+      } else {
+        bterm_descriptor->highlight(pin->getPinAsBTerm(), painter);
+      }
+
+      auto* source_node = pin->getPairedNode();
+      if (source_node == nullptr) {
+        continue;
+      }
+
+      const odb::Rect source_rect = source_node->getPinBBox();
+      const odb::Rect sink_rect = pin->getPinBBox();
+
+      const odb::Point source_pt(0.5 * (source_rect.xMin() + source_rect.xMax()),
+                                 0.5 * (source_rect.yMin() + source_rect.yMax()));
+      const odb::Point sink_pt(0.5 * (sink_rect.xMin() + sink_rect.xMax()),
+                               0.5 * (sink_rect.yMin() + sink_rect.yMax()));
+
+      painter.setPen(color, true, line_width);
+      painter.drawLine(source_pt.x(), source_pt.y(), sink_pt.x(), sink_pt.y());
+    }
+  }
+
+  // annotate with depth
+  const auto text_anchor = gui::Painter::Anchor::CENTER;
+  const double text_margin = 0.8;
+  painter.setPen(gui::Painter::white, true);
+  for (const auto& [level, pins] : map_) {
+    for (const auto& pin : pins) {
+      const odb::Rect pin_rect = pin->getPinBBox();
+      const odb::Point pin_pt(0.5 * (pin_rect.xMin() + pin_rect.xMax()),
+                              0.5 * (pin_rect.yMin() + pin_rect.yMax()));
+
+      const std::string text = std::to_string(level);
+      const odb::Rect text_bound = painter.stringBoundaries(pin_pt.x(), pin_pt.y(), text_anchor, text);
+
+      if (text_bound.dx() < text_margin * pin_rect.dx() && text_bound.dy() < text_margin * pin_rect.dy()) {
```

Comment:
The pin rect can be pretty small - for example the right pin gets no text unless I zoom in further:
![image](https://user-images.githubusercontent.com/761514/146799750-bb269b86-b29d-4ce6-9fbf-ea459b421637.png)

Reply:
I made the text scale 2x instead of 0.8. We can play around with that if needed.

Reply:
Using getPinLargestBox leads to a rect with large area but narrow width in this case:
![image](https://user-images.githubusercontent.com/761514/147601745-cd07db95-ebe8-41ed-bba1-a36534f77763.png)
It might be better to have widest pin box (perhaps above some height threshold) instead.